### PR TITLE
Sa 3067 get jc system app

### PR DESCRIPTION
**Pull request by kmaranionjc**
## Issues
* [SA-3067](https://jumpcloud.atlassian.net/browse/sa-3067) - SA-3067-Get-JCSystemApp

## What does this solve?
Enables admins to search software installed in their systems
## Is there anything particularly tricky?
Not really, if `-search `is not used filters are case sensitive especially `-softwareName`. 
-search can be time consuming since it searches all the systems in the org
## How should this be tested?
Run:
`Get-JCSystemApp` - returns all software in all systems
`Get-JCSystemApp -systemId "YOURID"` - returns all software installed in the specific system
`Get-JCSystemApp -systemOs "Windows"` - also test MacOs and Linux, this should return all apps in specific OS
`Get-JCSystemApp -systemOs "Windows" -softwareName "Microsoft Edge"`
`Get-JCSystemApp -systemOs "Windows" -softwareName "Microsoft Edge" -version "108.0.1462.42"`
Repeat the last 3 but instead of `-systemOs` use `-systemId`
`Get-JCSystemApp -softwareName "disk utility" -search`
`Get-JCSystemApp -softwareName "disk utility"  -systemId -search`

## Screenshots

image
image

[SA-3067]: https://jumpcloud.atlassian.net/browse/SA-3067?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+external help file: JumpCloud-help.xml
+Module Name: JumpCloud
+online version: https://github.com/TheJumpCloud/support/wiki/
+schema: 2.0.0
+---
+
+# Get-JCSystemApp
+
+## SYNOPSIS
+Returns the System Apps installed to a JumpCloud System(s)
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:11]:
```suggestion
Returns the applications/programs/linux packages installed on JumpCloud managed system(s). This function queries separate system insights tables to get data for macOS/windows/linux devices.
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+```
+
+### Search
+```
+Get-JCSystemApp [-SystemID <String>] [-SystemOS <String>] [-SoftwareName <String>] [-SoftwareVersion <String>]
+ [-Search] [<CommonParameters>]
+```
+
+## DESCRIPTION
+Get-JCSystem app function enables admins to search what software is on their systems
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:28]:
```suggestion
Get-JCSystem app function helps admins identify what applications/programs or linux packages exist on their JumpCloud managed systems.
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\> Get-JCSystemApp -SystemId 'SystemId'
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:34]:
```suggestion
PS C:\> Get-JCSystemApp -SystemId '6363237ec991136ae59892e4'
```

File: PowerShell/JumpCloud Module/Public/Systems/Get-JCSystemApp.ps1 [hunk 4]
```diff
@@ -0,0 +1,337 @@
+function Get-JCSystemApp () {
+    [CmdletBinding(DefaultParameterSetName = 'All')]
+    param (
+
+        [Parameter(Mandatory = $false, HelpMessage = 'The System Id of the system you want to search for applications')][ValidateNotNullorEmpty()]
+        [string]$SystemID,
+        [Parameter(Mandatory = $false , ValueFromPipelineByPropertyName, HelpMessage = 'The type (windows, mac, linux) of the JumpCloud Command you wish to search ex. (Windows, Mac, Linux))')]
+        [ValidateSet('Windows', 'MacOs', 'Linux')][ValidateNotNullorEmpty()]
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Public/Systems/Get-JCSystemApp.ps1:8]:
```suggestion
        [ValidateSet('Windows', 'macOs', 'Linux')][ValidateNotNullorEmpty()]
```

Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Public/Systems/Get-JCSystemApp.ps1:8]: According to apple macOS has a lowercase 'm' and capitalized 'OS'

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+```
+
+Returns the applications installed in the system with the given -SystemId
+
+### Example 2
+```powershell
+PS C:\> Get-JCSystemApp -SystemOs 'MacOs'
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:41]:
```suggestion
PS C:\> Get-JCSystemApp -SystemOs 'macOS'
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+```
+
+Returns the 'MacOs' systems and all the applications installed for each system
+
+### Example 3
+```powershell
+PS C:\> Get-JCSystemApp -SystemOs 'MacOs' -SoftwareName 'JumpCloud-Agent'
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:48]:
```suggestion
PS C:\> Get-JCSystemApp -SystemOs 'macOS' -SoftwareName 'JumpCloud-Agent'
```

Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:48]: This does not return anything as an example

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+```
+
+### Example 4
+```powershell
+PS C:\> Get-JCSystemApp -SystemOs 'MacOs' -SoftwareName 'JumpCloud-Agent' -SoftwareVersion '1.12.5'
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:53]:
```suggestion
PS C:\> Get-JCSystemApp -SystemOs 'macOS' -SoftwareName 'JumpCloud-Agent' -SoftwareVersion '1.12.5'
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+```
+
+Returns the applications installed in the system with the given -SystemId
+
+### Example 2
+```powershell
+PS C:\> Get-JCSystemApp -SystemOs 'MacOs'
+```
+
+Returns the 'MacOs' systems and all the applications installed for each system
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:44]:
```suggestion
Returns the 'macOS' systems and all the applications installed for each system
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+```powershell
+PS C:\> Get-JCSystemApp -SystemOs 'MacOs' -SoftwareName 'JumpCloud-Agent'
+```
+
+### Example 4
+```powershell
+PS C:\> Get-JCSystemApp -SystemOs 'MacOs' -SoftwareName 'JumpCloud-Agent' -SoftwareVersion '1.12.5'
+```
+
+Returns the 'MacOs' systems that have a 'JumpCloud Agent' application with the version '1.12.5'
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:56]:
```suggestion
Returns the 'macOS' systems that have a 'JumpCloud Agent' application with the version '1.12.5'
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+### Example 5
+```powershell
+PS C:\> Get-JCSystemApp -SoftwareName 'jumpcloud-agent' -Search
+```
+
+Returns any 'jumpcloud-agent' software installed in all the os systems
+
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:64]: I think we should also call out search by os and id examples here. Another example which would be immedietly helpful for folks would be:

Get-JCSoftwareApp -SystemOS "windows" 

Which would just return all windows programs

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+## PARAMETERS
+
+### -Search
+Global search ex.
+(1.1.2)
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:69]: parameter does not take input

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: False
+Accept wildcard characters: False
+```
+
+### -SystemOS
+The type (windows, mac, linux) of the JumpCloud Command you wish to search ex.
+(Windows, Mac, Linux))
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:132]:
```suggestion
(Windows, macOS, Linux))
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+```yaml
+Type: System.String
+Parameter Sets: (All)
+Aliases:
+Accepted values: Windows, MacOs, Linux
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:138]:
```suggestion
Accepted values: Windows, macOS, Linux
```

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: False
+Accept wildcard characters: False
+```
+
+### -SoftwareName
+The name of the application you want to search for ex.
+(JumpCloud-Agent, Slack)
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:85]: This is a decision point all of us need to make before we move forward...

for macOS Systems: `softwareName` refers to the `bundle_name` but we can technically refer to any field in the SI table. Not only do we need to decide which "name" parameter we want to query but we need to document this  so customers know what value will technically return valid results for `softwareName` parameters on windows/ macOS/ linux.


Screen Shot 2023-01-26 at 10 50 43 AM

Comment (gweinjc) [reply PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:85]: Based off of what Joe said, I think the `BundleExecutable` or `Name` field would be the likely candidate for finding the desired software app

**Comment by jworkmanjc** [comment]
Run:
Get-JCSystemApp - returns all software in all systems
Get-JCSystemApp -systemId "YOURID" - returns all software installed in the specific system ✅ (returns same amount of data as SDK)
Get-JCSystemApp -systemOs "Windows" - also test MacOs and Linux, this should return all apps in specific OS ✅
Get-JCSystemApp -systemOs "Windows" -softwareName "Microsoft Edge" ✅
Get-JCSystemApp -systemOs "Windows" -softwareName "Microsoft Edge" -**softwareversion** "108.0.1462.42" ✅
Repeat the last 3 but instead of -systemOs use -systemId
Get-JCSystemApp -softwareName "disk utility" -search ✅
Get-JCSystemApp -softwareName "disk utility"  -systemId -search ✅ yep returns more results then I'm looking for but helps me figure out what I'm looking for
Screen Shot 2023-01-26 at 1 54 03 PM

File: PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md (hunk 14 above)
Comment (kmaranionjc) [reply PowerShell/JumpCloud Module/Docs/Get-JCSystemApp.md:85]: Decided with filtering with property `name` on the macOs

**Review by jworkmanjc — CHANGES_REQUESTED** [review]
I'm game to merge this into the release branch assuming we address the issues raised.

**Review by gweinjc — APPROVED** [review]

**Pull request merged**